tests/unit/azure/test_credentials.py

```diff
@@ -0,0 +1,385 @@
+                ValidationResult(operation=None, result=ValidationResultResult.FAIL, message="fail"),
+            ],
+        )
+    if read_only:
+        return ValidateStorageCredentialResponse(
+            is_dir=True,
+            results=[ValidationResult(operation=ValidationResultOperation.READ, result=ValidationResultResult.PASS)],
+        )
+    return ValidateStorageCredentialResponse(
+        is_dir=True,
+        results=[ValidationResult(operation=ValidationResultOperation.WRITE, result=ValidationResultResult.PASS)],
+    )
+
+
+@pytest.fixture
+def credential_manager(ws):
+    ws.storage_credentials.list.return_value = [
+        StorageCredentialInfo(aws_iam_role=AwsIamRole("arn:aws:iam::123456789012:role/example-role-name")),
+        StorageCredentialInfo(
+            azure_managed_identity=AzureManagedIdentity("/subscriptions/.../providers/Microsoft.Databricks/...")
+        ),
+        StorageCredentialInfo(
+            name="included_test",
+            azure_service_principal=AzureServicePrincipal(
+                "62e43d7d-df53-4c64-86ed-c2c1a3ac60c3",
+                "b6420590-5e1c-4426-8950-a94cbe9b6115",
+                "secret",
+            ),
+        ),
+        StorageCredentialInfo(azure_service_principal=AzureServicePrincipal("directory_id_1", "app_secret2", "secret")),
+    ]
+
+    ws.storage_credentials.create.side_effect = side_effect_create_storage_credential
+    ws.storage_credentials.validate.side_effect = side_effect_validate_storage_credential
+
+    return StorageCredentialManager(ws)
+
+
+def test_list_storage_credentials(credential_manager):
+    assert credential_manager.list() == {"b6420590-5e1c-4426-8950-a94cbe9b6115", "app_secret2"}
+
+
+def test_list_included_storage_credentials(credential_manager):
+    include_names = {"included_test"}
+    assert credential_manager.list(include_names) == {"b6420590-5e1c-4426-8950-a94cbe9b6115"}
+
+
+def test_create_storage_credentials(credential_manager):
+    sp_1 = ServicePrincipalMigrationInfo(
+        StoragePermissionMapping(
+            "prefix1",
+            "app_secret1",
+            "principal_write",
+            "WRITE_FILES",
+            "directory_id_1",
+        ),
+        "test",
+    )
+    sp_2 = ServicePrincipalMigrationInfo(
+        StoragePermissionMapping(
+            "prefix2",
+            "app_secret2",
+            "principal_read",
+            "READ_FILES",
+            "directory_id_1",
+        ),
+        "test",
+    )
+
+    storage_credential = credential_manager.create_with_client_secret(sp_1)
+    assert sp_1.permission_mapping.principal == storage_credential.name
+    assert storage_credential.read_only is False
+
+    storage_credential = credential_manager.create_with_client_secret(sp_2)
+    assert sp_2.permission_mapping.principal == storage_credential.name
+    assert storage_credential.read_only is True
+
+
+def test_validate_storage_credentials(credential_manager):
+    service_principal = MagicMock()
```

Comment:
Why do you need to mock if we have a dataclass?

Reply:
I use mock just to save lines of code populating those dataclass fields that are not actually used. In this case, only the privilege field is needed.

Reply:
It's incorrect. You mock behavior, not data. Dataclasses are data. Mocking them is a bug, because these tests won't catch a bug anymore when someone else changes your code